Complete Documentation

emilgoldsmith opened the pull request:
Our project has become so large in terms of lines and code / number of files, and also in terms of how many (new) developers we have working together, so we require high quality documentation in order for this project to live on. This is what I try to provide here.

I'd really love it if you guys could give me a thorough code review, both in content but also grammar etc. as clarity of writing makes a big difference in how helpful the documentation is.

Look forward to having this in our project!

emilgoldsmith commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,28 @@
+<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
+<!--- If you're unsure about any of these, don't hesitate to ask. -->
+
+- [ ] My code follows the code style of this project. (see the [style guide](./docs/the-gazelle-style-guide.md))
```
I'm not sure if this link will work within a PR template, didn't think I could test it, but just realized I can probably just try and edit this into my PR message and play with it to see if it works.

emilgoldsmith commented on docs/the-gazelle-style-guide.md:
```diff
@@ -0,0 +1,136 @@
+# The Gazelle's Programming Style Guide
```
A thing I could maybe add here is references to examples in our codebase showing how to do things right. I think it's already pretty good though, and I don't know if we have really amazing examples yet?

emilgoldsmith commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,28 @@
+<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
+<!--- If you're unsure about any of these, don't hesitate to ask. -->
+
+- [ ] My code follows the code style of this project. (see the [style guide](../tree/master/docs/the-gazelle-style-guide.md))
```
I tested this link by editing my PR message, and this will work (when it's merged into master)

emilgoldsmith commented:
Just pushed a pretty significant update as I realized I had missed a few things, if you had already looked and want to check just the changes see here: https://github.com/thegazelle-ad/gazelle-server/pull/419/files/da18ff7..f809ece

emilgoldsmith commented:
Okay then... I couldn't help myself and documented our directory structure haha https://github.com/thegazelle-ad/gazelle-server/pull/419/files/f809ece634b81757db82733d271b44e85c5db92b..1b14f49f72351e0f783e7d7f0b3ab0db0959657d#diff-3f26975f3b01109ec051c12794a52722

navyasuri commented on docs/guides/dependency-learning-resources.md:
```diff
@@ -0,0 +1,95 @@
+- [a nice article on asynchronous functions from the perspective of Promises](https://medium.com/@bluepnume/learn-about-promises-before-you-start-using-async-await-eb148164a9c8)
+- [an async/await tutorial](https://codeburst.io/javascript-es-2017-learn-async-await-by-example-48acc58bad65)
+- [a good article on bad habits in async/await](https://medium.freecodecamp.org/avoiding-the-async-await-hell-c77a0fb71c4c)
+- [an async await tutorial](https://hackernoon.com/6-reasons-why-javascripts-async-await-blows-promises-away-tutorial-c7ec10518dd9)
```
Not the best article imho. I feel that the ones above are enough

navyasuri commented on docs/guides/dependency-learning-resources.md:
```diff
@@ -0,0 +1,95 @@
+- [an async/await tutorial](https://codeburst.io/javascript-es-2017-learn-async-await-by-example-48acc58bad65)
+- [a good article on bad habits in async/await](https://medium.freecodecamp.org/avoiding-the-async-await-hell-c77a0fb71c4c)
+- [an async await tutorial](https://hackernoon.com/6-reasons-why-javascripts-async-await-blows-promises-away-tutorial-c7ec10518dd9)
+
```
Once done understanding Promises, I think an article or two + just looking at code examples is good enough to understand async/await. So here might be an awesome place to link couple of great examples of async/await code.

emilgoldsmith commented on docs/guides/dependency-learning-resources.md:
```diff
@@ -0,0 +1,95 @@
+
+> NOTE TO REVIEWER FROM EMIL: I barely read any of the above yet, as I quickly put this together this morning as I remembered I should, will try to read them later and actually check for quality and relevance, they have lots of upvotes on Medium though and look quite good. If you want to read them and help me out feel free ;).
+
+We also use a lot of the newer ES6 and above syntax, a good guide for the most common ones that we basically use all of see this one:
```
Note to self, fix typo, you should probably add a 'for' after the comma

emilgoldsmith commented:
Yeah thanks @navyasuri ! That was really helpful. I added in one of your resources that I really liked and deleted some of the less helpful ones for Promises / async/await. In terms of videos, yeah of course people learn differently, so feel free to add some videos to the documentation later. I personally prefer written content myself as I feel it's faster.

If you can approve it (or give more comments) I think I'll just merge it. Also it's just documentation so it's not as important that it's perfect(as compared to code that might break in production) as it can be iterated on

navyasuri commented:
I think it's good to go!
The best part about documentation is we can always add to it when we find something that could be useful to everone else!

emilgoldsmith commented:
> The best part about documentation is we can always add to it when we find something that could be useful to everone else!

Yes exactly, and I hope to encourage a culture of doing that, especially with the PR template.

@navyasuri could you just officially "approve" it through the PR interface? Then I don't need to use my admin privileges to override our Github settings to merge it :)
